### ISSUE TITLE
Tolerance for type missmatch in _FillValue

### ISSUE THREAD

**Issue by ChrisBarker-NOAA** [Unidata/netcdf-c#3038]
I recently had a user provide a file that failed with my code, because we were trying to edit a netcdf file in place, and the lib couldn't open the file for writing.

It turns out the problem was this:

```
	float uo(time, depth, latitude, longitude) ;
		string uo:units = "m s-1" ;
		uo:_FillValue = NaN ;
		string uo:standard_name = "eastward_sea_water_velocity" ;
		string uo:long_name = "Eastward velocity" ;
```
note that the _FillValue does not have a type, and the `NaN`, as I understand it, is double `NaN`, whereas the variable is a `float`. (so that should be NaNf, I think). this type mismatch raises an error with nccopy:

```
NetCDF: Not a valid data type or _FillValue type mismatch
Location: file ?; fcn ? line 1408
```

NOTE: similar, but less informative error if I try to open it with the Python netCDF4 lib, in mode `r+` or `a`

So: yes, this is a bug in the file, and I'm going to try to reach out to the producers of the file, so they can correct. But this, and others, may be in archives for who knows how long!

How this file got created, I don't know, but perhaps older versions of the lib could let people make this mistake.

It was suggested on the mailing list that the library could be more tolerant, and at least allow, for instance, an nccopy, and maybe editing of any other variable, simply passing through the "error".

And, indeed, it could be really smart, and detect that if the *value* of the `_FillValue` can be represented in the array's type, than it could allow it -- and/or even change the type of the attribute on write.

Examples:
- You can represent NaN in either double or a float, you can represent -999 in most integer types, etc.

- But you can't represent NaN in an integer type, or a value beyond the range of that type.

I know this is tricky in C, but maybe worth it?

Thanks,

Tested on Version 4.9.2 -- confirmed by multiple other folks on I'm not sure which recent versions...

Tested on Windows, OS-X and Linux

**Comment by Dave-Allured** [comment]
Here is a reproducer for the malformed file.  This is not a minimal reproducer because it demonstrates only that a recent version of NCO Utilities can GENERATE a new file with the same malformation.  Start with Chris's original malformed file, attached to this mailing list message:

https://www.unidata.ucar.edu/mailing_lists/archives/netcdfgroup/2024/msg00019.html

Then simply this, with simplified file name:

`ncks orig.nc copy.nc`

The _FillValue data type does not match the file variable.  This is discussed later in the same list thread.  This works for me on Mac with NCO 5.2.1, netcdf 4.9.2, and hdf5 1.14.3.

**Comment by DennisHeimbigner** [comment]
Sigh! This again.  See Issues https://github.com/Unidata/netcdf-c/issues/1766 and https://github.com/Unidata/netcdf-c/issues/1299.
We should just kill this constraint and modify the documentation (except what to to do about the formal
netcdf specifications that talk about this constraint).

**Comment by ChrisBarker-NOAA** [comment]
Can we appeal to Postel's law: "be conservative in what you send, be liberal in what you accept".

So the lib should certainly never create a file from scratch that is inconsistent with _FillValue types.

But it could accept one (which it does if it's read-only).

The challenge is what to do when opening an existing file for writing *r+, a" -- 

Would it be possible to change the type of the _FillValue attribute to match the variable?

**Comment by Dave-Allured** [comment]
The original complaint is failure to write to an existing netcdf-4 file.  The evidence so far points to aggressive guards in third party softwares, and NOT in the netcdf-C library.  All things considered, especially Dennis's historical perspective, I think that current behavior of netcdf-C is reasonable and DOES allow such writes.

Unless someone can demonstrate an actual fault in netcdf-C, I recommend that this issue be closed and reposted to third party softwares; netcdf4-python in particular, perhaps also NCO.  For the record, NCL gets the same failure, but has not been maintainable for the past several years.

This contradicts Chris's request that the lib should never allow _FillValue with inconsistent type to be created.  I suggest that mismatch should be considered acceptable but discouraged, which appears to be the current situation.

**Comment by ChrisBarker-NOAA** [comment]
I'm pretty sure the error is coming from the C lib, not the Python wrapper. It appears immediately when you try your open the file.

And nccopy also fails. 

This is likely a feature, rather than bug, but it does appear to be in the C lib. 

I'm not set up to try working with the C lib directly-- but could test this out if need be.

**Comment by Dave-Allured** [comment]
Chris, agreed, the nccopy failure is interesting, and it is part of the netcdf-C package.  However it is a separate utility, not part of the C library itself.

The fact that ncatted can write to one mismatched file (copy.nc) is strong circumstantial evidence that the netcdf-C library is working correctly with _FillValue mismatch.  Your original error posted to the mailing list, plus my experience with inconsistent ncatted behavior, points to a second and so far undiagnosed problem with your original file.  Now *that* is interesting.  Someone needs to look at that more carefully before deciding that there is something wrong with the netcdf-C library.

What happens if you try writing to copy.nc with your netcdf4-python library?  I mean the original copy.nc from above, before "fixing" the mismatch with ncatted.

**Comment by ChrisBarker-NOAA** [comment]
A few more tests:

```
$ nccopy bad_fill_value.nc bad_fill_value_copied.nc 
NetCDF: Not a valid data type or _FillValue type mismatch
Location: file ?; fcn ? line 1408
```
it does generate a tiny (239 bytes) file:
```
$ ncdump bad_fill_value_copied.nc 
netcdf bad_fill_value_copied {
}
```
so presumably opened a new file, and then barfed out when it tried to start writing to it.

NCO, however, is able to copy the file, preserving the _FIllValue data type missmatch. The resulting file behaves the same (i.e. doesn't work) with nccopy and netCDF4.

NCO is using libnetcdf, so apparently it's possible for the lib to write such invalid files, if called a certain way.

"something wrong with the netcdf-C library." -- I don't think it's wring, but maybe not optimum?

@czender: any ideas who NCO works, and nccopy doesn't for this malformed file?

**Comment by DennisHeimbigner** [comment]
So I did some experiments about this.

First, it is possible to create files with bad _FillValue with the netcdf-c library. What you do is the following steps:
    1. create the file and variable of type float, but do not try to set the fillvalue for the variable.
    2. create an attribute of type double and with name, say, "FV".
    3. rename the FV attribute to "_FillValue".
 
This appears to produce a malformed file similar to the one under discussion. See the attached program t.c.txt
[t.c.txt](https://github.com/user-attachments/files/17264336/t.c.txt)


Second, it possible to repair the bad file as follows:
1. open the bad file in writeable mode using the NC_WRITE flag to nc_open().
2. read the _FillValue attribute to get its value (a double in this case).
3. convert that value to equivalent float value
4. write a new attribute, FV say, of type NC_FLOAT and with the float value constructed in step 3.
5. delete the _FillValue attribute
6. rename the FV attribute to _FillValue

See attached file tfix.c.txt
[tfix.c.txt](https://github.com/user-attachments/files/17264331/tfix.c.txt)

**Comment by DennisHeimbigner** [comment]
A note: t.c produces a file called "badfile.nc", while tfix.c modifies a file called "fixfill.nc".
You create fixfill.nc by "cp badfile.nc fixfill.nc". I did this to preserve the original badfile.c.

**Comment by czender** [comment]
Yes, that sound a lot like what NCO does, and I would not be surprised if ncatted could therefore be used to "fix" the malformed files.

**Comment by ChrisBarker-NOAA** [comment]
> open the bad file in writeable mode using the NC_WRITE flag to nc_open().

Hmm -- with the Python lib, the 'w' mode will overwrite the file, so that's not helpful.
the 'a' and 'r+' modes are apparently the same, and both result in an error:
```
OSError                                   Traceback (most recent call last)
Cell In[7], line 1
----> 1 ds = netCDF4.Dataset("bad_fill_value.nc", 'r+')

File src/netCDF4/_netCDF4.pyx:2469, in netCDF4._netCDF4.Dataset.__init__()

File src/netCDF4/_netCDF4.pyx:2028, in netCDF4._netCDF4._ensure_nc_success()

OSError: [Errno -101] NetCDF: HDF error: 'bad_fill_value.nc'
```

So apparently I can't get the NC_WRITE flag from the Python interface :-(

which is odd, because I jsut looked at the code, and it *should* be NC_WRITE:
```
cmode = NC_WRITE if append_mode else NC_NOWRITE
...
ierr = nc_open(path, cmode, &grpid)
```
(there's a lot of special casing in there, so *maybe* I missed sometihng, but I don't think so.

So perhaps the issue is:
```
_ensure_nc_success(ierr, err_cls=OSError, filename=path)
```
which raises a Python exception if the open call returned an error:

    if ierr == NC_NOERR:
        return
<process the error and raise an exception>
```
my C is weak -- but it looks like in your fixing code you are proceeding even if there is an error from the open() call?

In any case, we now know how to "fix" a file with either NCO or the C lib, but I think the question at hand is whether the C lib should do something else to "smooth over" these errors. 

Maybe this is too much magic, but it *could* potentially:

1) (the too much magic option) When opening a file with NC_WRITE, and it detects a type miss match with _FillValue, it goes through the above hacky steps to correct the _FillValue type -- returning an error only if the value of _FillValue is not losslessly convertible to the correct type.

2) When opening a file with NC_WRITE, don't check the _FillValue type -- and let the incorrect file pass on through. This is what it used to do, yes?

It still seems l like a good idea for the lib to not let you create a missmatch when starting a file from scratch.

**Comment by edwardhartnett** [comment]
I agree with option 2 - when opening the file, do not check the fill value type.

**Comment by DennisHeimbigner** [comment]
The nc_open does not fail, which is, in retrospect, surprising.
I suspect the reason it succeeds is because netcdf-c uses lazy read for attributes.
So when you actually tried to read the _FillValue, it would cause a failure.

**Comment by DennisHeimbigner** [comment]
Yes, case 2 -- read-wide/write-narrow (mentioned above) -- is basically the right choice..

I think that when a mismatch occurs, that the _FillValue should be converted internally to the same type as the containing variable. So in the above example, the nan value would be converted to nanf. Conversion between numeric and char types are normally illegal in netcdf-c, but we would probably have to allow it in this case, which means we need to define conversion rules.

There is a minor question about reading the fillvalue. This can be done in two ways.
1.  using nc_inq_var_fill
2. using nc_get_att asking for the value of _FillValue attribute.

The question is: under what circumstances should the original value be returned vs the converted internal value?
I would propose that nc_inq_var_fill actually returns the converted value. But nc_get_att would return the original, un-converted value.

**Comment by ChrisBarker-NOAA** [comment]
> The nc_open does not fail, which is, in retrospect, surprising.

hmm -- but it does fail in the python lib, which wouldn't raise if nc_open didn't return an error code. -- mysterious...

> I think that when a mismatch occurs, that the _FillValue should be converted internally to the same type as the containing variable. 

+1

> Conversion between numeric and char types are normally illegal in netcdf-c, but we would probably have to allow it in this case, which means we need to define conversion rules.

Or have an error in that case. because you can convert type if the value is compatible, e.g. NaN and NaNf, or a 999 and a signed or unsigned int of various sizes. 

But there is no way to know what's intended if the _FillValue is, e.g. "MM", and the type of the array is numeric -- *maybe* you could choose a NaN for a float type, but for an integer type? who knows?

Or for a _FillValue of -1 for a unsigned integer array -- what could you do?

> The question is: under what circumstances should the original value be returned vs the converted internal value?
I would propose that nc_inq_var_fill actually returns the converted value. But nc_get_att would return the original, un-converted value.

Is that even possible? once it's been converted, wouldn't the attribute value and type be changed? In any case, I think that keeping them consistent should be the goal. It would get really messy if they weren't.

**Comment by Dave-Allured** [comment]
A simple C program shows that the current netcdf-C 4.9.2 library can easily write to a netcdf file with a _FillValue mismatch.  See attachment [write-test1.c.txt](https://github.com/user-attachments/files/17270444/write-test1.c.txt).  It is necessary to first "sanitize" the user's original data file with ncks.  As expected, exactly three data values were overwritten:
```
> ncks bad_fill_value.nc bad_fill_value_copied.nc
> gcc -g -O0 -Wall -I/opt/local/include -L/opt/local/lib -lnetcdf write-test1.c
> ./a.out
> nccmp -dNsf bad_fill_value.nc bad_fill_value_copied.nc 
DIFFER : VARIABLE : uo : POSITION : [0,0,0,0] : VALUES : 0.027 <> 10
DIFFER : VARIABLE : uo : POSITION : [0,0,0,1] : VALUES : 0.037 <> 20
DIFFER : VARIABLE : uo : POSITION : [0,0,0,2] : VALUES : 0.05 <> 30
```
Repeating this test on the ORIGINAL file:
```
> gcc -g -O0 -Wall -I/opt/local/include -L/opt/local/lib -lnetcdf write-test2.c
> ./a.out
Error: write-test2.c, line 29:
Netcdf error #-103, NetCDF: Can't write file
File = bad_fill_value.nc
```
That error message is not very informative.  Considering everyone's reports so far, I see two distinct problems:

1.  The guards preventing writes with _FillValue mismatch are in external software such as the netcdf4-python library, but NOT in the netcdf-C library.
2.  There is a second, as yet undiagnosed problem in the user's original file, `bad_fill_value.nc`.  This problem is blocking writes from ncatted, pure netcdf-C, and probably everything else.

The netcdf-C library is already satisfying the original request, "Tolerance for type missmatch in _FillValue".  Also take into account the historical reasons mentioned by Dennis.  Therefore I recommend the following.

1.  Make no further changes to current _FillValue handling in the netcdf-C library.
2.  Close this issue as satisfied, and "won't fix".
3.  Follow up with a correctly targeted issue to netcdf4-python.
4.  Only if someone has the time, investigate the second problem in the user's original file.  To me this seems to be an obscure corner case, therefore low priority.

**Comment by Dave-Allured** [comment]
... And there it is.  Behind door number 2.
```
        /* Without creation ordering, file must be read-only. */
        if (!h5->no_write)
            BAIL(NC_ECANTWRITE);
```
So either user's original file was made by a really ancient version of HDF5, or else somebody somewhere went through all the right hoops to disable creation order tracking.  You have 17+ years of history to dig through, to find out how *that* happened.

**Comment by ChrisBarker-NOAA** [comment]
> The guards preventing writes with _FillValue mismatch are in external software such as the netcdf4-python library, but NOT in the netcdf-C library.

I don't think that's the case, I did look at the netcdf4-python code, and all it does is raise an exception IF the call to nc_open results in an arror code -- and it then reports the error from the C lib.

NOTE: I confirmed these same results using teh netCDF4Python lib -- after passing through ncks, it can be opened for writing. I'm a bit confused now, though, 'cause I get that same uninformative error message, but I nkow i saw something about _FillValue type mismatch somewhere -- where?? Looking at my oriingla pst, I got this error message:
```
NetCDF: Not a valid data type or _FillValue type mismatch
Location: file ?; fcn ? line 1408
```
which I'm not getting now -- curious -- maybe a different netcdf version? I"ll have to try to figure that out later.

> There is a second, as yet undiagnosed problem in the user's original file, bad_fill_value.nc. This problem is blocking writes from ncatted, pure netcdf-C, and probably everything else.

That is a possibility, I suppose. That'll take some digging :-( it does seem that ncks "fizes" the file, without changing the _FillValue type so what does it change?

Question: `nc_inq_var_fill`
apparently gives a different result than checking the _FillValue attriube -- is that value stored in the netCDF file? could that be what ncks changes?

I think the original user got that file from Copernicus, and it's model results from 2023-09-01 -- so it couldn't have been written more than 1 year or so ago. I seriously doubt anyone went through any machinations to mess the file up, and I'd be surprised if their HDF version was THAT old, but ?????

**Comment by Dave-Allured** [comment]
> ```
> NetCDF: Not a valid data type or _FillValue type mismatch
> Location: file ?; fcn ? line 1408
> ```
> which I'm not getting now -- curious -- maybe a different netcdf version?

That message is from running nccopy on the user's original file.  Nccopy may have been responding to either one of the two out-of-spec conditions.

 > it does seem that ncks "fixes" the file, without changing the _FillValue type so what does it change?

Ncks rewrites the entire file with modern versions of the netcdf and HDF5 libraries, and conforming format options.  As long as ncks (NCO) can *read* the data from any input file, it can also write a cleaned-up copy, with modern and conforming format details.

> Question: `nc_inq_var_fill` apparently gives a different result than checking the _FillValue attriube -- is that value stored in the netCDF file? could that be what ncks changes?

Right.  For any file variable that has an added fill value, there are actually two copies stored.  One is the netcdf version, stored as an actual attribute.  As others have explained, this is what can allow a data type mismatch.  The other fill value is baked into the variable's HDF5 internal structure, and is type bound.

> I think the original user got that file from Copernicus, and it's model results from 2023-09-01 -- so it couldn't have been written more than 1 year or so ago. I seriously doubt anyone went through any machinations to mess the file up, and I'd be surprised if their HDF version was THAT old, but ?????

Good point.  Now my best guess that the original file creators were using a third party netcdf/HDF5 writer that was lacking in certain conformance details.  Creation order tracking can be switched on or off, even in modern HDF5 versions.  It is "supposed to be" always enabled to write conforming netcdf files.

**Comment by Dave-Allured** [comment]
Chris, can you please close this issue now?  It is proved that the netcdf-C library is tolerant of _FillValue mismatches, and will allow writes to such files.  Postel's law is satisfied.  All other related failures are now explained, except for the exact cause of that particular netcdf4-python error.

If you need the netcdf4-python library to tolerate _FillValue mismatches, please open a ticket there.

It does not seem to me that it is worth pursuing the creation order problem with netcdf-C, python libraries, or the original data provider.  This is a rare case.  Anyone who wants to modify this out-of-spec file can easily create a sanitized copy and operate on the copy.

**Comment by ChrisBarker-NOAA** [comment]
Closing as won't fix:

Not worth the effort to make the netcdflib be able to interact directly with this particular type of malformed file.

**Comment by Dave-Allured** [comment]
Thanks for closing the ticket.

> > that the netcdf-C library is tolerant of _FillValue mismatches, and will allow writes to such files.
> 
> Well, it depends on what you mean by mismatch -- I _think_ the problem I brought up is apparently a missmatch between the internal HDF fill value and the array data type ...

No, that is not even possible in the HDF5 file structure.  There is no separate data type stored for the internal HDF5 fill value.   [HDF5 format spec 3.0, paragraphs IV.A.2.e and IV.A.2.f](https://support.hdfgroup.org/documentation/hdf5/latest/_f_m_t3.html#title3), fill value messages:  **"The fill value is interpreted with the same datatype as the dataset."**

> But it also the case that it would be possible for the netcdf lib to tolerate that particular malformed file, using a similar hack to what NCO uses.

Not possible.  As already shown above, the core netcdf-C lib is blocked by the out of spec creation order, not by any fill value problem.  The original error that you posted to netcdfgroup was this.  Error -103 is creation order error, not fill value:
```
OSError: [Errno -103] NetCDF: Can't write file: 'cmems_mod_nws_phy_anfc_0.027deg-3D_P1D-m_1727700325546.nc'
```
Fill value got dragged into this because of secondary testing and speculation.  The original problem is not a fill value problem.

> As to how rare it is -- I'm not sure -- it seems this issue has been brought up before, and unfortunately, Copernicus is a pretty popular data source :-(

Copernicus should not be publishing netcdf files that are that far out of spec.  On the other hand, you can certainly *read* the file.  Why is your server trying to modify original source files?  Are you adding attributes, or something?  For a case like this, I would say make a file copy, or else insert a special case hack into the OpenDAP server.

**Comment by ChrisBarker-NOAA** [comment]
Thanks for the clarification.

> Fill value got dragged into this because of secondary testing and speculation.

nccopy is what raised the `_FillValue` error:

```
$ nccopy bad_fill_value.nc junk.nc
NetCDF: Not a valid data type or _FillValue type mismatch
Location: file ?; fcn ? line 1408
```

Which is why I headed down that path. And, indeed the _FillValue was a different type than the array. Coincidence?

I wonder why it reported that error?

Really beating a dead horse now, but since I got curious, I tried opening that file for writing with h5py, which wraps:
hdf5  - 1.14.3

It could open it for writing, and write to successfully.

The result was a file that still couldn't be copied by nccopy but the copy command gave a different error:

NetCDF: HDF error
Location: file ?; fcn ? line 2016

So ???

And it gets stranger:

If I open it for writing with h5py, and change the _FillValue to the correct type, then nccopy works.

But the copied file still won't open for writing with netCDF4.

So it appears that it WAS a coincidence -- the error keeping it from being opened with nc_write was:

Error -103: creation order error
 
But the error keeping nccopy from working was the _FillValue type missmatch. Two different issues.

I'm still confused why h5py can tolerate the creation order issue, as the netcdf lib is linked to the same version as hdf5 -- but at least part of the mystery is solved.

**Comment by ChrisBarker-NOAA** [comment]
> Copernicus should not be publishing netcdf files that are that far out of spec.

We certainly agree on that -- but apparently most people only read them, and that works, so they haven't realized they have a problem.

> On the other hand, you can certainly read the file. Why is your server trying to modify original source files? Are you adding attributes, or something?

Altering data in this case -- shifting timezones and/or changing the longitude reference frame.

> For a case like this, I would say make a file copy

Well, a regular file system copy won't work, nor nccopy -- but NCO does, so that may be our solution, if we bother making one -- we can tell our users to fix their files first :-)

**Comment by guigrpa** [comment]
Lead developer of the [MyOcean data portal](https://data.marine.copernicus.eu/viewer/expert) here, where the offending file comes from. I wanted to add some information that is missing from this thread.

* The data portal uses the stock h5py library to create netCDF files.
* Thanks to a user reporting to Copernicus Marine User Support and linking to this thread, we've been alerted to the badly-typed `_FillValue`. Instead of `NaN`, it should have been `NaNf`. This has been fixed in the portal.

Please [reach out to us](https://help.marine.copernicus.eu/en/) if you still encounter difficulties downloading data using the MyOcean data portal.

**Comment by Dave-Allured** [comment]
@guigrpa Thank you for responding.  The user's original file from MyOcean data also contained an out-of-spec creation order error, as discussed above.  How did that happen?  What version of stock h5py created that file?  Has that error condition been fixed in the portal?

**Comment by Dave-Allured** [comment]
> Could you please provide more details about this creation-order issue? Thanks for any pointer that can lead us in the right direction!

Certainly.  Let us see if the creation order error still persists after your current fix.  The original bad file was "cmems_mod_nws_phy_anfc_0.027deg-3D_P1D-m_1727700325546.nc", [as reported above](https://github.com/Unidata/netcdf-c/issues/3038#issuecomment-2392389028).  Can you please provide me with the current version of that data file, so that I may check it and demonstrate the error?  Please attach the file to this github issue.

**Comment by Dave-Allured** [comment]
Also please show me how this file is listed in the MyOcean data catalog.  What is the URL for the catalog?

**Comment by guigrpa** [comment]
Maybe you need a multi-variable netCDF to demonstrate the creation-order issue.

Here is an example from exactly the same dataset as the original bad file in the thread, with both East and North sea water velocity.

[cmems_mod_nws_phy_anfc_0.027deg-3D_P1D-m_1730915253706.nc.zip](https://github.com/user-attachments/files/17651042/cmems_mod_nws_phy_anfc_0.027deg-3D_P1D-m_1730915253706.nc.zip)

And the URL where it was obtained from: https://data.marine.copernicus.eu/-/pce477be71

**Comment by guigrpa** [comment]
> Also please show me how this file is listed in the MyOcean data catalog. What is the URL for the catalog?

https://data.marine.copernicus.eu/viewer/expert?view=catalogue

**Comment by Dave-Allured** [comment]
@guigrpa The creation order problem still exists in the new data file that you just sent.  Here is a simple reproducer: [write-test3.c.zip](https://github.com/user-attachments/files/17651541/write-test3.c.zip).  This test program simply tries to overwrite a few values in the main data variable.  This is with the netcdf-c 4.9.2 library, the current release.
```
> ls -go *.nc *.c
-rw-r--r--  1   83368 Nov  6 10:42 cmems_mod_glo_phy-thetao_anfc_0.083deg_P1D-m_1730914963583.nc
-rw-r--r--  1    1296 Nov  6 11:18 write-test3.c

> gcc -g -O0 -Wall write-test3.c -I/opt/local/include -L/opt/local/lib -lnetcdf
> ./a.out
Error: write-test3.c, line 29:
Netcdf error #-103, NetCDF: Can't write file
File = cmems_mod_glo_phy-thetao_anfc_0.083deg_P1D-m_1730914963583.nc
```

**Comment by Dave-Allured** [comment]
Now as I said earlier, error -103 is creation order error, not fill value.  This error is **not** explained well in netCDF documentation.  If you trace this error number in netcdf-c source code, you will eventually get to the creation order check [that I described above](https://github.com/Unidata/netcdf-c/issues/3038#issuecomment-2395561013).

The creation order requirement for netCDF-4 is explained [here in the version 2 (draft) NetCDF Users Guide](https://docs.unidata.ucar.edu/nug/2.0-draft/nc4_file_format.html#creation_order).

This episode is now fully explained.  Copernicus and MyOcean are trying to generate netCDF-4 files using `h5py`.  That is an **HDF5** library, not a **netCDF** library.  Therefore you are technically generating **HDF5** files, not fully conforming **netCDF-4** files.  You may continue to fine tune your `h5py` driver to improve conformance.

However, netCDF-4 includes several augmentations to the basic HDF5 format.  It might also be a moving target.  Therefore I think you should consider switching to one of the official **netCDF-4 libraries** for ongoing **netCDF-4** file generation.

**Comment by ChrisBarker-NOAA** [comment]
@Dave-Allured: thanks for this -- it wasn't at all clear to me what a "creation order" error meant. 

In short, am I correct in saying that these files are "correct" HDF5 files, but not correct netCDF4 files?

**Comment by dopplershift** [comment]
@guigrpa Are you all using h5py directly, or is it through h5netcdf?

**Comment by guigrpa** [comment]
> That is an HDF5 library, not a netCDF library. Therefore you are technically generating HDF5 files, not fully conforming netCDF-4 files. You may continue to fine tune your h5py driver to improve conformance.

For context, we're using h5py because netCDF4 was not working properly on Pyodide (Python environment for browser-side execution) when the functionality was originally implemented. We will periodically review the ecosystem and switch to netCDF4 as soon as it's supported in that environment.

Thanks for the inputs regarding the creation-order issue. We'll look into it and post our results here.

**Comment by guigrpa** [comment]
@Dave-Allured We've just published an update to the [MyOcean data portal](https://data.marine.copernicus.eu/) that enables creation-order tracking. Here is an example file created with the new version, in case you may want to check:

[cmems_mod_glo_phy-cur_anfc_0.083deg_P1D-m_1731063952602.nc.zip](https://github.com/user-attachments/files/17677643/cmems_mod_glo_phy-cur_anfc_0.083deg_P1D-m_1731063952602.nc.zip)

Unfortunately, I cannot run your reproducer easily in my setup, but we did some tests with `h5dump -H -p -q creation_order ...` that confirmed that creation-order is correctly enabled.

**Comment by ChrisBarker-NOAA** [comment]
I do not have @Dave-Allured's expertise to know for sure, but I can confirmt hat that file. can be opened in append mode,  and written too -- so it looks good to me -- thanks!

**Comment by Dave-Allured** [comment]
> [cmems_mod_glo_phy-cur_anfc_0.083deg_P1D-m_1731063952602.nc.zip](https://github.com/user-attachments/files/17677643/cmems_mod_glo_phy-cur_anfc_0.083deg_P1D-m_1731063952602.nc.zip)

@guigrpa Thanks for the quick portal update!  Yes I also confirm that your new file version is configured correctly.  It passes my creation order and write test.

I have another conformance request.  This is purely an informational issue, not critical.  Since you are using a non-standard netcdf-4 driver, please add the **_NCProperties global attribute** to identify your driver configuration.  This will help with future troubleshooting.  This special attribute is described [here in the version 2 (draft) NetCDF Users Guide](https://docs.unidata.ucar.edu/nug/2.0-draft/nug_conventions.html#NCProperties).

Here is a current example using `ncdump -hs` on a normal netcdf-4 file:
```
:_NCProperties = "version=2,netcdf=4.9.2,hdf5=1.14.5"
```
Please use the version 2 layout for conformance.  For MyOceanfiles, there should be at least three components for full identification, something like this.  Use your official driver name(s) in place of the dummy name that I inserted:
```
:_NCProperties = "version=2,myoceanwriter=9.9.9,h5py=9.9.9,hdf5=1.14.5"
```
Thanks for your help with this.
